src/main/scala/mesosphere/marathon/core/storage/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -0,0 +1,142 @@
+
+/**
+  * A Write Ahead Cache of another persistence store that lazily loads values into the cache.
+  *
+  * @param store The store to cache
+  * @param mat a materializer for Akka Streaming
+  * @param ctx The execution context for future chaining.
+  * @tparam K The persistence store's primary key type
+  * @tparam Serialized The serialized format for the persistence store.
+  */
+class LazyCachingPersistenceStore[K, Category, Serialized](
+    store: BasePersistenceStore[K, Category, Serialized])(implicit
+  mat: ActorMaterializer,
+    ctx: ExecutionContext) extends PersistenceStore[K, Category, Serialized] with StrictLogging {
+
+  private[storage] val idCache = RwLock(TrieMap[Category, Seq[Any]]())
+  private[storage] val valueCache = RwLock(TrieMap[K, Option[Any]]())
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = {
+    val storeId = ir.category
+    idCache.read { ids =>
+      if (ids.contains(storeId)) {
+        Source(ids(storeId).map(_.asInstanceOf[Id]))
+      } else {
+        val childFuture = async {
+          val children = await(store.ids.toMat(Sink.seq)(Keep.right).run())
+          idCache.write(c => c(storeId) = children)
+          children
+        }
+        Source.fromFuture(childFuture).mapConcat(identity)
+      }
+    }
+  }
+
+  override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    async {
+      await(store.deleteAll(k))
+      val storageId = ir.toStorageId(k, None)
+      valueCache.write { vc =>
+        vc.remove(storageId)
+        idCache.write { c =>
+          val old = c.getOrElse(ir.category, Nil)
+          val children = old.filter(_ != storageId)
+          if (children.nonEmpty) {
+            c.put(ir.category, children)
+          } else {
+            c.remove(ir.category)
+          }
+        }
+      }
+      Done
+    }
+  }
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = {
+    val storageId = ir.toStorageId(id, None)
+    val cached = valueCache.read(_.get(storageId))
+    cached match {
+      case Some(v) =>
+        Future.successful(v.asInstanceOf[Option[V]])
+      case None =>
+        async {
+          val value = await(store.get(id))
+          valueCache.write(_.put(storageId, value))
+          value
+        }
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] =
+    store.get(id, version)
+
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    async {
+      await(store.store(id, v))
+      val storageId = ir.toStorageId(id, None)
+      val storageCategory = ir.category
+      valueCache.write { values =>
```

Comment:
I think this will not guarantee the correct order of updates in the cache, since the `store` operation is performed asynchronously. In the EntityStoreCache we introduced an update closure, that was called inside the future:
https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala#L51